soc: nxp: Consolidate nxp soc configs and code

decsny opened the pull request:
NXP SOCs seem to be getting a lot of copy pasted code in different places, try to consolidate it, as well as some other minor fixups and restructures.

Commit:
soc: nxp: Move kinetis pinctrl to include/

There are some SOCs which use the same pinctrl
device, but are not really overall that similar to the kinetis
family of SOCs, so prepare to reuse this code by other SOC families.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Do not change driver init priorities

Doesn't seem like there is any reason to be changing the
serial and ADC driver init priorities in the SOC defconfigs.
By default, clock control has priority 30, dma has priority 40,
and serial/adc have priority 50. This seems already to be the order
that is needed, and the original goal of these overrides, which
don't seem necessary anymore.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: mcx: Fix Kconfig files and MFD config

The Kconfig defaults should be in Kconfig.defconfig.

MFD should not be made dependent on the LP flexcomm being
enabled, since MFD can be for other things than just flexcomm.
Change from depends on to an if.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Remove RT1061 and RT1051 refs

NXP is not supporting these derivative parts in the zephyr upstream,
these references to them should therefore be removed in order to avoid
confusion.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Move flexspi log level change to driver

Single point of control over this kconfig's effect.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: mcx: Do not use family level config

Move all dependencies of the family config to series level,
and put a disclaimer saying not to use the family config.

Change all occurrences of the family config in code to the
MCXNX4X series config.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Centralize flexspi related configuration

Currently this code related to how to configure the
flash size and address when using flexspi to XIP is copy
pasted in all sort of places and ways all over the tree,
let's clean this up and have single point of control over
this configuration.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: rw: Delete unused configs

Delete some unused kconfigs in RW SOC family.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
drivers: usb device: Remove MCUX controller type

The CONFIG_USB_MCUX_CONTROLLER_TYPE choice config
is not necessary, let's just remove it. Theoretically,
if there was an SOC with the EHCI and LPCIP3511, then
it might cause a build error, but there is no SOC
which is supported on this legacy driver that has that.

Remove all settings of it in the SOC files.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Consolidate ROM RAMLOADER config

Consolidate the ROM RAMLOADER config to be in one place,
add a new Kconfig file to be included by the SOCs with the feature.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: nxp: Move connectivity defaults to common

Move default of monolithic and fw loader, options, etc to the
common kconfig files rather than defaulting in soc file.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: rw: Consolidate clock cycle kconfig.

No need for this to be expanded out so much.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

mmahadevan108 commented on soc/nxp/kinetis/Kconfig.defconfig:
```diff
@@ -3,14 +3,6 @@
-	default 55
-	depends on SERIAL
-
-config ADC_INIT_PRIORITY
```
This does not seem to be related to this commit, should we not change this priority?

decsny replied:
it doesn't seem necessary, since it looks like ADC already initializes after DMA

dleach02 replied:
I suspect this has the same reasoning as the SERIAL.

dleach02 replied:
But digging into history and it looks like the initial PR from 2020 set it to this value

decsny replied:
updated the commit message

dleach02 commented on soc/nxp/kinetis/Kconfig.defconfig:
```diff
@@ -3,14 +3,6 @@
 
-config SERIAL_INIT_PRIORITY
-	default 55
-	depends on SERIAL
```
The root of the override of SERIAL is from this PR #39452 to fix a problem that someone found with the order of initialization between SERIAL and the CLOCK control drivers. The root problem was that both classes of drivers had the same priority and by a somewhat undefined linker behavior the initialization order just happened to work until someone built with a different build environment and the linker decided to order things in a reverse order. The referenced PR created the generic config for SERIAL and with a separate commit the NXP parts were forced to this priority.

We would need to make sure that this doesn't get the same initialization priority as the clock control driver before removing this. 

I also note that I had a draft PR that was pulled in favor of the above PR and we appear to have missed pulling over the comment block that was above this code explaining the reasoning for the override.

decsny replied:
the clock control has priority 30 by default, serial has priority 50 by default. so changing it to 55 doesn't seem necessary.

dleach02 replied:
At the time of the change they were in the same priority. I think there is additional activity in the project to break this type of coding of priority and leverage something in device tree to configure dependencies.

decsny replied:
the init priorities are a topic in Arch WG right now, as far as this PR, just want to remove these overrides which are no longer relevant

axelnxp commented on modules/Kconfig.mcux:
```diff
@@ -366,18 +366,21 @@ config HAS_MCUX_XBARA
 
 config NXP_FW_LOADER
 	bool "Include firmware loader component"
+	default y if (BT || WIFI)
```
The firmware loader is not used on all NXP BLE platforms, currently only RW uses it.
MCXW71 loads its firmware through the ROM, using nxp blhost tool.
So, I think we should keep it disabled by default and enable it in the soc defconfig.
Or maybe `default y if SOCXXX` but this could become kinda bloated ?

axelnxp commented on modules/Kconfig.mcux:
```diff
@@ -366,18 +366,21 @@ config HAS_MCUX_XBARA
 
 config NXP_RF_IMU
 	bool "Include RF_IMU adapter for intercore messaging"
+	default y if (BT || WIFI)
```
The RF_IMU component is specific to RW, won't be used on the MCXW family

decsny replied:
okay, then should to config be introduced in the RW/Kconfig rather than in this file?

decsny commented:
* Updated commit message regarding driver priority override removal
* Moved some BT defaults back to RW soc kconfig
* Fixed check compliance errors (except for one)

Commit:
soc: nxp: imxrt: Clang format flexspi_nor_config.h

Check compliance is not allowing changes to this file
without running clang-format.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

decsny commented:
Fixed an accidental squash

decsny commented:
fixed the last check compliance error (a clang format check error) by disabling clang format around the offending area of code, because clang format was introducing another compliance error when run

@pdgendt looks like you added this check of clang format to check compliance, want to make you aware of this problem now where apparently clang format does not align with the other compliance checks

decsny commented:
> The clang format checks are informative, and do not block merging the PR

it was blocking CI on the PR until I added this comment to except those lines of code

the reason I pinged you was in case you didn't know there was this conflict between clang format and the other checks

Commit:
drivers: flash/memc: Source logging kconfig last

Some flash/memc drivers like flexspi will want to default the
value of the log level to off to avoid RWW hazard while XIP,
to do this, the logging template must be sourced after the driver
kconfig files so that the default value from the driver is able
to be checked, since logging template introduces an unconditional
default otherwise.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

Commit:
soc: imxrt118x: Cleaning up kconfigs

Moving RT118x kconfigs to series level instead of family level,
and cleaning up redundant declarations, and adding missing
HAS_MCUX_FLEXSPI selection.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

danieldegrasse commented on soc/nxp/imxrt/imxrt118x/Kconfig.defconfig:
```diff
@@ -23,12 +23,15 @@ config DCDC_VALUE
 	default 0x400 if BOOT_FLEXSPI_NOR
 
+config ROM_START_OFFSET
+	default 0xb000 if BOOT_FLEXSPI_NOR && SOC_SERIES_IMXRT118X
```
Isn't this already dependent on `SOC_SERIES_IMXRT118X`, since it is within the `if SOC_SERIES_IMXRT118X` block?

decsny replied:
yes, it was a copy paste, I'm not sure why you phrased it like a question, I'll just update it to remove it in case I need to update this PR again, otherwise not worth the CI rerun IMO

danieldegrasse commented on soc/nxp/mcx/Kconfig.defconfig:
```diff
@@ -5,4 +5,32 @@ if SOC_FAMILY_NXP_MCX
+endchoice
+
+# Code relocation is needed when FLASH_MCUX_FLEXSPI_XIP is enabled
+config CODE_DATA_RELOCATION_SRAM
```
This seems like it will turn on code relocation whenever `FLASH_MCUX_FLEXSPI_XIP` is enabled, but it looks like the MCX only actually uses code relocation when the FLEXSPI flash driver is enabled: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/mcx/mcxn/CMakeLists.txt#L16, so should this setting depend on `FLASH_MCUX_FLEXSPI_NOR`?

decsny replied:
no, it's common to all the SOC due to the drivers being relocated https://github.com/zephyrproject-rtos/zephyr/blob/0a7908588c755d3b8dfa3cc5c967d8a085762500/drivers/flash/CMakeLists.txt#L66

danieldegrasse replied:
Ok- it looks like you dropped this definition and moved to the common one, which does have the right dependencies (as far as I can tell)

danieldegrasse commented on drivers/flash/Kconfig:
```diff
@@ -86,10 +86,6 @@ menuconfig FLASH
 
 if FLASH
 
-module = FLASH
```
To make this change we have to deviate from the pattern present in other driver `Kconfig` files (I looked at the serial and SPI ones for example).

Could we instead disable logging at the SOC level, maybe in `soc/nxp/common/Kconfig.flexspi_xip`?

decsny replied:
no, I don't care about this pattern, it probably is done thoughtlessly anyways, specific drivers should be able to override driver class config defaults, the other classes just haven't run into the issue yet. I am not putting it in the SOC level because the default value we want has specifically to do with the driver, it belongs here

danieldegrasse replied:
I think other classes have run into this issue, and the solution was to define a Kconfig for that *driver's* log level, rather than using the subsystem Kconfig: https://github.com/zephyrproject-rtos/zephyr/blob/cfe5bd922432d9c2090e9cbf149edd7172dc4e62/drivers/gpio/Kconfig.fxl6408#L20 Perhaps we should do the same here, because I could see someone wanting debug logging to be enabled for their on chip flash driver, but disabled for the FlexSPI

danieldegrasse replied:
Looking at this more, some driver Kconfig files define the logging string later (such as I2C: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/Kconfig#L143) and some do not. So I think this approach is fine as is

danieldegrasse commented on soc/nxp/common/Kconfig.flexspi_xip:
```diff
@@ -0,0 +1,41 @@
+	  FlexSpi when planning to execute code in FlexSpi Memory.
+
+config CODE_DATA_RELOCATION_SRAM
+	default y if FLASH_MCUX_FLEXSPI_XIP && MEMC_MCUX_FLEXSPI
```
I *think* the dependency here is on the flash driver, not on the MEMC driver. For example on the RT595 we can use the memc driver while running in XIP mode to interface with HyperRAM on the second FlexSPI device, and we don't need to relocate code to SRAM

decsny replied:
this is what the dependency was before this PR, I just moved it to a different file, while you might be right, I don't want to attempt this change in this PR

danieldegrasse commented on soc/nxp/common/Kconfig.rom_loader:
Commit message needs to be updated, the definition for this Kconfig is not being added to the common flexspi kconfig file

decsny replied:
true, I had originally put it in the other file, I guess I forgot to update the message

danieldegrasse reviewed (commented):
Very exciting cleanup, just a few additional comments

DerekSnell reviewed (commented):
Hi @decsny ,
I don't see an issue with these changes.  However, I want more eyes on this, and don't want my review to count as +1.

Thanks

decsny commented on soc/nxp/common/Kconfig.flexspi_xip:
```diff
@@ -0,0 +1,41 @@
+# SPDX-License-Identifier: Apache-2.0
+
+DT_CHOSEN_Z_FLASH := zephyr,flash
+DT_COMPAT_FLEXSPI := nxp,imx-flexspi
```
thanks but not worried about it right now, i'm sure it can be handled when it's added

EmilioCBen commented on soc/nxp/rw/Kconfig.defconfig:
```diff
@@ -25,27 +26,15 @@ endif # CORTEX_M_SYSTICK
 if BT
 
-config FLASH
+config NXP_MONOLITHIC_BT
```
Where is this used?

manuargue commented on include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:
```diff
@@ -0,0 +1,73 @@
+#define PORT_PCR_SRE_MASK 0x0
+#endif
+
+#define Z_PINCTRL_KINETIS_PINCFG(node_id)                                                          \
```
Beware that S32K1 is very similar (peripheral wise) to Kinetis KE1 series and both are using the same pinctrl driver. `Z_PINCTRL_KINETIS_PINCFG` is defined `zephyr/soc/nxp/s32/s32k1/pinctrl_soc.h`.

I don't think it's okay to move this header here, because even the hw IP is the same, each soc has an specific integration

dleach02 replied:
@manuargue do you have a recommendation on how this one should be handled? A more generic common name perhaps?

manuargue replied:
I'm not requesting a change in the naming, although that's something we will need to do eventually but is a bigger change involving bindings, etc. My request on this particular change is to keep the pinctrl headers for Kinetis under `soc/nxp/kinetis/common/pinctrl_soc.h`. I think the whole point of the `pinctrl_soc.h` header is to be in the soc folder, hence the suffix `_soc` :)

decsny replied:
@manuargue completely disagree with this and this is a crucial part of the PR which is important for NPI we are trying to enable which will use this file that is not in the kinetis family, we don't want it in two places. This doesn't even affect s32 if it doesn't choose to include this header in it's pinctrl_soc.h,  kinetis still has it's own pinctrl_soc.h which is just including this file

decsny commented on include/zephyr/drivers/pinctrl/pinctrl_soc_kinetis_common.h:
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2022, 2024 NXP
```
yeah, I just copy pasted the whole file, I suppose I can take this opportunity to delete the (c)

axelnxp commented on drivers/bluetooth/hci/Kconfig.nxp:
```diff
@@ -25,6 +25,21 @@ config HCI_NXP_SET_CAL_DATA_ANNEX100
 
 if BT_H4_NXP_CTLR
 
+config FLASH
```
I don't think those config should be set under `BT_H4_NXP_CTLR`.
`BT_H4_NXP_CTLR` is not selected on RW, `BT_NXP` is.
`BT_H4_NXP_CTLR` is for host platform that will communicate with a tranceiver through the uart h4 driver.

Lyle from Tony's team should review this too, @dleach02 do you know Lyle's tag on github ?

axelnxp replied:
Yes I agree

axelnxp reviewed (approved):
Changes related to BT look good to me, thanks for the clean up
